File: 프로그래머스/2/131130. 혼자 놀기의 달인/혼자 놀기의 달인.py
def solution(cards):
    visited = [False for _ in range(len(cards))]
    groups = findGroup(cards, visited)
    if len(groups) == 1:
        return 0
    groups.sort(reverse=True)
    return groups[0] * groups[1]

def findGroup(cards, visited):
    groups = []
    for i in range(len(cards)):
        if not visited[i]:
            visited[i] = True
            idx = cards[i] - 1
            count = 1
            while not visited[idx]:
                visited[idx] = True
                count += 1
                idx = cards[idx] - 1
            groups.append(count)
    return groups
        
            
    
        

File: 프로그래머스/2/131130. 혼자 놀기의 달인/혼자 놀기의 달인.java
import java.util.*;

class Solution {
    public int solution(int[] cards) {
        int answer = 0;
        for (int i = 0; i < cards.length; i++) {
            Map<Integer, Boolean> group1 = findGroup(cards, i);
            int max = 0;
            
            for (int j = 0; j < cards.length; j++) {
                if (group1.containsKey(j)) {
                    continue;
                }
                
                int groupSize = findGroup(cards, j).size();
                max = (groupSize > max) ? groupSize : max;
            }
            
            answer = (group1.size() * max > answer) ? group1.size() * max : answer;
        }
        return answer;
    }
    
    public Map<Integer, Boolean> findGroup(int[] cards, int N) {
        Map<Integer, Boolean> group = new HashMap<>();
        boolean isNotChecked = true;
        int next = N;    
        
        while (isNotChecked) {
            if (group.containsKey(next)) {
                break;   
            }
            group.put(next, true);
            next = cards[next] - 1;
        }
        
        return group;
    }
}